bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -634,6 +639,21 @@ public Bookie(ServerConfiguration conf, StatsLogger statsLogger)
         this.indexDirsManager = createIndexDirsManager(conf, diskChecker, statsLogger.scope(LD_INDEX_SCOPE),
                                                        this.ledgerDirsManager);
 
+        this.allocator = ByteBufAllocatorBuilder.create()
+                .poolingPolicy(conf.getAllocatorPoolingPolicy())
+                .poolingConcurrency(conf.getAllocatorPoolingConcurrency())
+                .outOfMemoryPolicy(conf.getAllocatorOutOfMemoryPolicy())
+                .outOfMemoryListener((ex) -> {
+                    try {
+                        LOG.error("Unable to allocate memory, exiting bookie", ex);
+                        shutdown(-1);
+                    } finally {
+                        Runtime.getRuntime().halt(-1);
```

Comment:
Don't call this directly here.  Bookie is used in all tests, and halt getting called in the middle of a test is a pain to track down (especially since in OOM scenario, the LOG may fail).

Reply:
How should be pass it? 

What I wanted to achieve here is: 
 * Try to shutdown bookie gracefully (mainly to close ZK session such that new ledgers won't choose this bookie)
 * No matter what, bring the process down after that attempt. This would protect from a thread hanging on in background to keep JVM alive.

Reply:
@merlimat there is an `uncaughtExceptionHandler` in bookie server (https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieServer.java#L78).

There are two approaches:

1) pass the uncaught exception handler down from bookie server to bookie. so that bookie can trigger uncaught exception handler when OOM

2) construct the allocator outside of bookie instance. so when the allocator is constructed in bookie server, it can use the uncaught exception handler when OOM